fix: deep merge behavior in flat config

fasttime opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

Fixes #17820

**Tell us about your environment (`npx eslint --env-info`):**

Node version: v21.5.0
npm version: v10.2.4
Local ESLint version: v8.56.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 23.1.0

**What parser are you using (place an "X" next to just one item)?**

[x] `Default (Espree)`
[ ] `@typescript-eslint/parser`
[ ] `@babel/eslint-parser`
[ ] `vue-eslint-parser`
[ ] `@angular-eslint/template-parser`
[ ] `Other`

**Please show your full configuration:**

No config file.

**What did you do? Please include the actual source code causing the issue.**

I compared the merge behavior in flat config against eslintrc in different cases.

```js
// test1.js
const { default: { FlatESLint } } = await import("eslint/use-at-your-own-risk");

const eslint = new FlatESLint({
    overrideConfigFile: true,
    baseConfig: [
        {
            languageOptions: {
                parserOptions: {
                    foo: {}
                }
            },
        },
        {
            languageOptions: {
                parserOptions: {
                    // null overwrites other values in eslintrc.
                    foo: null
                }
            },
        },
    ]
});
const { languageOptions: { parserOptions } } = await eslint.calculateConfigForFile("file.js");
console.log(parserOptions);
```

```js
// test2.js
const { default: { FlatESLint } } = await import("eslint/use-at-your-own-risk");

const eslint = new FlatESLint({
    overrideConfigFile: true,
    baseConfig: [
        {
            languageOptions: {
                parserOptions: {
                    prop1: {},
                    prop2: {},
                    prop3: { 1: "one" },
                    prop4: ["a", "b", "c"],
                    prop5: { ["__proto__"]: 1 }
                }
            },
        },
        {
            languageOptions: {
                parserOptions: {
                    // In eslintrc, primitives overwrite other values.
                    prop1: 123,

                    // In eslintrc, strings are not spread.
                    prop2: "foo",

                    // In eslintrc, it is possible to merge an object or an array with another array.
                    // Properties in the second array override properties in the first value.
                    prop3: ["zero"],
                    prop4: ["d", "e"],

                    // An own property "__proto__" is not merged in eslintrc.
                    prop5: { ["__proto__"]: 2 }
                }
            },
        },
    ]
});
const { languageOptions: { parserOptions } } = await eslint.calculateConfigForFile("file.js");
console.log(parserOptions);
```

**What did you expect to happen?**

The merge behavior in flat config should be the same as in eslintrc.

```console
❯ node test1.js
{ foo: null }
```

```console
❯ node test2.js
{
  prop1: 123,
  prop2: 'foo',
  prop3: [ 'zero', 'one' ],
  prop4: [ 'd', 'e', 'c' ],
  prop5: {}
}
```

**What actually happened? Please include the actual, raw output from ESLint.**

There are some differences in the merge behavior, all addressed in this PR.

```console
❯ node test1.js
Error: Key "languageOptions": Key "parserOptions": Cannot convert undefined or null to object
```

```console
❯ node test2.js
{
  prop1: {},
  prop2: { '0': 'f', '1': 'o', '2': 'o' },
  prop3: [ 'zero' ],
  prop4: [ 'd', 'e' ],
  prop5: { ['__proto__']: 2 }
}
```

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**Repro**: https://stackblitz.com/edit/stackblitz-starters-7ta5nx

#### What changes did you make? (Give an overview)

Updated the `deepMerge` algorithm in lib/config/flat-config-schema.js to make it more similar to eslintrc's [`mergeWithoutOverwrite`](https://github.com/eslint/eslintrc/blob/bdce01aad747393d13cd1ae17d8280e54c9c5191/lib/config-array/config-array.js#L126-L156). I think this was requested in https://github.com/eslint/eslint/issues/17820#issuecomment-1843181566.

The new changes affect the way properties are merged: we could keep all of them or revert some to the current behavior.

* Primitive property values in the second object will always overwrite properties of the first object. Currently, if the first property value is an object and the second one is a primitive, the algorithm will try to merge the values, producing a new object that is mostly, but not always, a clone of the first one.
* `null` property values in the second object will no longer result in a runtime error but instead behave like other primitives and overwrite properties of the first object. This is the original bug reported in #17820.
* String property values in the second object will behave like other primitives and no longer spread over their characters.
* When merging two objects with an own `__proto__` property, the `__proto__` property will no longer be merged. In the current merge implementation, the result will contain an own `__proto__` property if the first object does.
* <strike>Arrays can merge with arrays and other objects like in eslintrc. The resulting value will be an array or a plain object, depending on the type of the second value. Currently, merging only occurs when the second value is not an array (although the first value may be any object). I'm not 100% sure of this change because of [this comment](https://github.com/eslint/eslint/blob/v8.56.0/lib/config/flat-config-schema.js#L73-L76).</strike> There is agreement that in the new config system arrays should not merge with other arrays or objects like in eslintrc.

#### Is there anything you'd like reviewers to focus on?

All of the above points.

As a hint, you may find it helpful to play with the [repro](https://stackblitz.com/edit/stackblitz-starters-7ta5nx) when not sure about how things work in the current implementation or in eslintrc.

<!-- markdownlint-disable-file MD004 -->

fasttime commented on tests/lib/config/flat-config-schema.js:
```diff
@@ -18,36 +40,14 @@ describe("merge", () => {
-
-        assert.strictEqual(result, second);
+        confirmLegacyMergeResult(first, second, result);
     });
```
I believe that these cases are not actually supported in eslintrc or in flat config, because arrays are never merged directly. So I've replaced them with similar unit tests below where arrays are wrapped in a property.

fasttime commented on tests/lib/config/flat-config-schema.js:
```diff
@@ -18,36 +40,14 @@ describe("merge", () => {
         const result = merge(first, second);
 
         assert.deepStrictEqual(result, { ...first, ...second });
-    });
-
-    it("overrides an object with an array", () => {
-        const first = { foo: 42 };
-        const second = ["bar", "baz"];
-        const result = merge(first, second);
-
-        assert.strictEqual(result, second);
-    });
-
-    it("merges an array with an object", () => {
-        const first = ["foo", "bar"];
-        const second = { baz: 42 };
-        const result = merge(first, second);
-
-        assert.deepStrictEqual(result, { 0: "foo", 1: "bar", baz: 42 });
-    });
-
-    it("overrides an array with another array", () => {
-        const first = ["foo", "bar"];
-        const second = ["baz", "qux"];
-        const result = merge(first, second);
-
-        assert.strictEqual(result, second);
```
I believe that these cases are not actually supported in eslintrc or in flat config, because arrays are never merged directly. So I've replaced them with similar unit tests below where arrays are wrapped in a property.

nzakas commented on lib/config/flat-config-schema.js:
```diff
@@ -98,16 +87,21 @@ function deepMerge(first, second = {}, mergeMap = new Map()) {
         ...second
     };
 
+    delete result.__proto__; // eslint-disable-line no-proto -- don't merge own property "__proto__"
```
If we don't want the `__proto__`, then I'd suggest using `Object.create(null)` to create `result`.

fasttime replied:
I think the intent here was just to exclude `__proto__` when it's an _own_ property, because the inherited property `Object.prototype.__proto__` should not be masked or accidentally overwritten (`delete` only removes own properties). By contrast, `Object.create(null)` creates an object with `null` prototype, that lacks all methods of `Object.prototype` but could still have an own property called `__proto__`, so it's not really the same thing.

fasttime replied:
Alternatively, we could treat `__proto__` as a regular property with the usual merge behavior as other properties. We'd just need to take care to use `Object.defineProperty` rather than a regular assignment (`=`) to set `__proto__` in order to not mess up with the prototype ([here](https://github.com/eslint/eslint/blob/8d84b0b94fa8a8b2d868d08ff20cf8df14a24dbb/lib/config/flat-config-schema.js#L119) and [here](https://github.com/eslint/eslint/blob/8d84b0b94fa8a8b2d868d08ff20cf8df14a24dbb/lib/config/flat-config-schema.js#L121)).

nzakas replied:
Ah gotcha, so this line is deleting an own property that might have been added through the spread operations. That makes sense to me. :+1:

nzakas reviewed (commented):
Thanks for doing this! I think that we should fix all of the eslintrc incompatibilities in this PR. It was definitely my intent to mirror eslintrc merging behavior. :+1:

mdjermanovic commented on tests/lib/config/flat-config-schema.js:
```diff
@@ -72,6 +83,39 @@ describe("merge", () => {
+        confirmLegacyMergeResult(first, second, result);
+    });
+
+    it("merges two arrays in a property", () => {
+        const first = { someProperty: ["foo", "bar", void 0, "baz"] };
+        const second = { someProperty: ["qux", void 0, 42] };
+        const result = merge(first, second);
+
+        assert.deepStrictEqual(result, { someProperty: ["qux", "bar", 42, "baz"] });
+        confirmLegacyMergeResult(first, second, result);
+    });
```
I think yes. I'd say the new config system is a good opportunity to change the confusing and possibly unintended specifics of eslintrc's deep merging, but would like to hear @nzakas's opinion too.

Merging array properties into objects and vice versa almost never makes sense.

Merging arrays into arrays might make sense in some specific cases, but I believe it's generally expected that arrays overwrite arrays.

So the behavior would be pretty simple to explain. Merging `first` with `second` results in:

* Deeply merged object if both `first` and `second` are plain objects (neither is an array).
* `first` if `second` is `undefined`.
* `second` in all other cases.

nzakas replied:
I think those changes make sense. Let's do it. 👍

fasttime replied:
Thanks! This should working as suggested after 5d435f6acb57c648bd1132fce287c163ce0281e2.

fasttime commented:
I've just noticed another edge case: when a property is undefined in both first and second object - but explicitly set in one or both - eslintrc does not set it at all in the result, while `deepMerge` would currently set it with value undefined.

```js
deepMerge({ someProperty: void 0 }, { someProperty: void 0 });
// current result:           { someProperty: undefined }
// merge result in eslintrc: {}
```

Any advice/preference on how to handle this?

mdjermanovic commented on lib/config/flat-config-schema.js:
```diff
@@ -98,16 +96,21 @@ function deepMerge(first, second = {}, mergeMap = new Map()) {
     };
 
+    delete result.__proto__; // eslint-disable-line no-proto -- don't merge own property "__proto__"
+    if (Array.isArray(second)) {
+        result = Object.assign([], result);
+    }
```
Looks like this can never happen, because `second` is always a non-array object?

fasttime replied:
Thanks! Removed in 6e2da74bf7c86b04347a742c8e702e87b9ee2644.

mdjermanovic commented:
> I've just noticed another edge case: when a property is undefined in both first and second object - but explicitly set in one or both - eslintrc does not set it at all in the result, while `deepMerge` would currently set it with value undefined.
> 
> ```js
> deepMerge({ someProperty: void 0 }, { someProperty: void 0 });
> // current result:           { someProperty: undefined }
> // merge result in eslintrc: {}
> ```
> 
> Any advice/preference on how to handle this?

I think it's fine to leave the property with `undefined` value in this case.

fasttime commented:
Added a test to check that undefined properties are actually set (b6732baba423e73547fe522f6a69c017e64dd06e). Also fixed an edge case to make sure that only own enumerable properties are set or merged. This matches the behavior in eslintrc.

mdjermanovic commented:
I've updated the title since we're not fully replicating eslintrc behavior.

mdjermanovic reviewed (approved):
LGTM, thanks! Leaving open for @nzakas to verify.

One thing that is different with this implementation, compared to eslintrc and the previous implementation, is that they were deep-cloning all objects and arrays, while this keeps the same instances except for objects that are getting merged with other objects in the process. However, I don't think this is important, as we're not modifying any objects or arrays that were passed in.

nzakas reviewed (approved):
LGTM. Thanks for all your work on this. 🙏
